### ISSUE TITLE
Spyk'ed object can't be  unmocked which leads to OutOfMemoryError

### ISSUE THREAD

**Issue by bakomchik** [mockk/mockk#1013]
### Expected Behavior
unmock(spyk()) -- works
No `java.lang.OutOfMemoryError: Java heap space` error occurred.

### Current Behavior

`java.lang.OutOfMemoryError: Java heap space` occurs

`unmock(spyk())` -- does not unmock spyk'ed object

#### Context

We have a lot of SpringBoot tests . 
We annotated `@Lazy` bean with `@SpykBean` in our TestCaseBase class
And sometimes new spring context created and bean get `spyked` again and it leads to OutOfMemoryError
because spiked object can't be unmocked/unspyked

* MockK version:     ` implementation("io.mockk:mockk:1.13.3")`
* OS:  `Darwin 21.6.0 Darwin Kernel Version 21.6.0: Thu Sep 29 20:13:56 PDT 2022; root:xnu-8020.240.7~1/RELEASE_ARM64_T6000 arm64`

* Kotlin version:  `1.7.21`
* JDK version: `openjdk 11.0.16.1 2022-08-12
OpenJDK Runtime Environment Homebrew (build 11.0.16.1+0)
OpenJDK 64-Bit Server VM Homebrew (build 11.0.16.1+0, mixed mode)
`
* Type of test: unit test

#### Minimal reproducible code (the gist of this issue)

```kotlin
// -----------------------[ GRADLE DEFINITIONS ] -----------------------
dependencies {
    testImplementation(kotlin("test"))
    implementation("io.mockk:mockk:1.13.3")
}

tasks.test {
    useJUnitPlatform()
}

tasks.withType<KotlinCompile> {
    kotlinOptions.jvmTarget = "1.8"
}

application {
    mainClass.set("MainKt")
}
// -----------------------[ YOUR CODE STARTS HERE ] -----------------------


fun main(args: Array<String>) {
    (0..1000).forEach {
        val spyk = spyk(LazySpringBeanWhichHoldsReferenceForBeanFactory(ByteArray(10 * 1024 * 1024)))
        every { spyk.doSmth() } returns "Wow"
        spyk.doSmth();
        clearMocks(spyk)
        unmockkObject(spyk)
    }
}

class LazySpringBeanWhichHoldsReferenceForBeanFactory(val beanFactory: ByteArray){
    fun doSmth():String{
            return "Smth"
    }
}
```

**Comment by bakomchik** [comment]
Path to gc roots
<img width="861" alt="Снимок экрана 2022-12-30 в 18 02 50" src="https://user-images.githubusercontent.com/3716745/210085762-99b7f6b5-20e5-4a7e-b38a-8ac4b1e9a2ca.png">
 The object kept by these strong references an can't be collected by gc
<img width="859" alt="image" src="https://user-images.githubusercontent.com/3716745/210085956-5e0fda72-e5be-4ee1-9fd0-ee9550bcaa27.png">

**Comment by bakomchik** [comment]
Looks like original object can't be collected by GC and removed from all maps  due to strong  reference via path
<img width="852" alt="Снимок экрана 2023-01-04 в 14 12 38" src="https://user-images.githubusercontent.com/3716745/210543167-d31936bd-7992-4fbb-9c27-9ce4758980ef.png">

**Comment by bakomchik** [comment]
I've created pull request https://github.com/mockk/mockk/pull/1018/files

**Comment by bakomchik** [comment]
I found that mock(proxy) objects ,that supposed to be collected by gc, can't be collected despite the fact that weak reference used for storing these objects in WeakMockHandlersMap, because of these proxy objects are strongly reachable via dispose routine

**Comment by mabwah-arista** [comment]
It seems the fix for #997 undid this, as now I'm seeing memory leaks due to spy objects not being GCed once cleared. There must be a way to allow these objects to be GCed when calling `clearAllMocks` while not having them disappear while they're still needed.
